x-pack/plugins/timelines/server/search_strategy/timeline/factory/helpers/format_timeline_data.ts

```diff
@@ -5,118 +5,116 @@
-    };
-  }
-  const formattedData = await getDataSafety(getDataFromFieldsHits, fieldToEval);
-  return formattedData.reduce((acc: TimelineNonEcsData[], { field, values }) => {
-    // nested fields return all field values, pick only the one we asked for
-    if (field.includes(fieldName)) {
-      acc.push({ field, value: values });
-    }
-    return acc;
-  }, []);
+  // Filter and map in one pass
+  return formattedData
+    .filter((item) => item.field.includes(fieldName))
+    .map(({ field, values }) => ({
+      field,
+      value: values,
+    }));
 };
 
-const mergeTimelineFieldsWithHit = async <T>(
-  fieldName: string,
-  flattenedFields: T,
-  hit: EventHit,
-  dataFields: readonly string[],
-  ecsFields: readonly string[]
-) => {
-  if (fieldName != null) {
-    const nestedParentPath = getNestedParentPath(fieldName, hit.fields);
-    if (
-      nestedParentPath != null ||
-      has(fieldName, hit.fields) ||
-      ECS_METADATA_FIELDS.includes(fieldName)
-    ) {
-      const objectWithProperty = {
-        node: {
-          ...get('node', flattenedFields),
-          data: dataFields.includes(fieldName)
-            ? [
-                ...get('node.data', flattenedFields),
-                ...(await getValuesFromFields(fieldName, hit, nestedParentPath)),
-              ]
-            : get('node.data', flattenedFields),
-          ecs: ecsFields.includes(fieldName)
-            ? {
-                ...get('node.ecs', flattenedFields),
-                ...buildObjectRecursive(fieldName, hit.fields),
-              }
-            : get('node.ecs', flattenedFields),
-        },
-      };
-      return merge(flattenedFields, objectWithProperty);
+export const formatTimelineData = async (
+  hits: EventHit[],
+  fieldRequested: readonly string[],
+  excludeEcsData: boolean
+): Promise<TimelineEdges[]> => {
+  const ecsFields = excludeEcsData ? [] : TIMELINE_EVENTS_FIELDS;
+
+  const uniqueFields = new Set([...ecsFields, ...fieldRequested]);
+  const dataFieldSet = new Set(fieldRequested);
+  const ecsFieldSet = new Set(ecsFields);
+
+  const results: TimelineEdges[] = new Array(hits.length);
+
+  for (let i = 0; i < hits.length; i++) {
+    const hit = hits[i];
+    if (hit._id) {
+      const result = createBaseTimelineEdges();
+
+      result.node._id = hit._id;
+      result.node._index = hit._index;
+      result.node.ecs._id = hit._id;
+      result.node.ecs.timestamp = getTimestamp(hit);
+      result.node.ecs._index = hit._index;
+
+      if (hit.sort?.length > 1) {
+        result.cursor.value = hit.sort[0];
+        result.cursor.tiebreaker = hit.sort[1];
+      }
+
+      result.node.data = [];
+
+      for (const fieldName of uniqueFields) {
+        const nestedParentPath = getNestedParentPath(fieldName, hit.fields);
+
+        if (
+          !nestedParentPath &&
+          !has(fieldName, hit.fields) &&
+          !ECS_METADATA_FIELDS.includes(fieldName)
```

Comment:
💅 

1. We could make `ECS_METADATA_FIELDS` with a `Set`, making the lookup even faster. It's only 4 items long at the moment, so the speed gains will be minimal.
2. Consider extracting  `ECS_METADATA_FIELDS.includes(fieldName)` to the beginning of the loop, so it only needs to be evaluated once.

Reply:
did 2, 1 I left as is as it is only 4 items and used elsewhere